databuilder/query_engines/base_sql.py

```diff
@@ -493,153 +470,163 @@ def round_to_first_of_month(self, date):
+        to be explicitly dropped or whether the database will discard it automatically
+        when the session terminates
+        """
+        raise NotImplementedError()
 
-    def get_aggregate_column(self, query, aggregate_node):
+    #
+    # DATABASE CONNECTION
+    #
+    @property
+    def engine(self):
+        if self._engine is None:
+            engine_url = sqlalchemy.engine.make_url(self.backend.database_url)
+            # Hardcode the specific SQLAlchemy dialect we want to use: this is the
+            # dialect the query engine will have been written for and tested with and we
+            # don't want to allow global config changes to alter this
+            engine_url._get_entrypoint = lambda: self.sqlalchemy_dialect
+            self._engine = sqlalchemy.create_engine(engine_url, future=True)
+            # The above relies on abusing SQLAlchemy internals so it's possible it will
+            # break in future -- we want to know immediately if it does
+            assert isinstance(self._engine.dialect, self.sqlalchemy_dialect)
+        return self._engine
+
+    def get_temp_table_name(self, name_hint):
         """
-        For an aggregate node, build the column to hold its value
-        Aggregate column names are a combination of column and aggregate function,
-        e.g. "patient_id_exists"
+        Return a table name based on `name_hint` suitable for use as a temporary table.
+
+        `name_hint` is arbitrary and is only present to make the resulting SQL slightly
+        more comprehensible when debugging.
         """
-        output_column = self.get_output_column_name(aggregate_node)
-        if aggregate_node.function == "exists":
-            return sqlalchemy.literal(True).label(output_column)
-        else:
-            # The aggregate node function is a string corresponding to an available
-            # sqlalchemy function (e.g. "exists", "count")
-            function = getattr(sqlalchemy.func, aggregate_node.function)
-            source_column = aggregate_node.column
-            return function(query.selected_columns[source_column]).label(output_column)
-
-    def apply_filter(self, query, filter_node):
-        # Get the base table
-        table_expr = get_primary_table(query)
-
-        column_name = filter_node.column
-        operator_name = filter_node.operator
-        # Does this filter require another table? i.e. is the filter value itself an
-        # Output node, which has a source that we may need to include here
-        value_expr, other_tables = self.get_value_expression(filter_node.value)
-        if other_tables:
+        self.temp_table_count += 1
+        return f"{self.temp_table_prefix}{name_hint}_{self.temp_table_count}"
+
+    def get_temp_database(self):
+        """Which schema/database should we write temporary tables to."""
+        return None
+
+
+def apply_filter(query, column, operator, value, value_query_node, or_null=False):
```

Comment:
No, because we don't want `sqlalchemy_utils` to know anything about the QM?

Reply:
Yes, that's right. This function has a bunch of weird behaviour that I haven't had time yet to think through and rework. My hope is that we'll eventually reduce it to some simple sqlalchemy operations at which point it can go in `sqlalchemy_utils`. But I'm very keen to keep `sqlalchemy_utils` ignorant of the rest of the codebase.